python/cuda_parallel/cuda/parallel/experimental/_iterators.py

```diff
@@ -55,12 +35,76 @@ def _ctypes_type_given_numba_type(ntype):
     return mapping[ntype]
 
 
-CompileResult = collections.namedtuple('CompileResult', ['ltoir', 'result_type'])
+@lru_cache(maxsize=256)  # TODO: what's a reasonable value?
+def cached_compile(func, sig, abi_name=None, **kwargs):
+    return numba.cuda.compile(func, sig, abi_info={"abi_name": abi_name}, **kwargs)
+
+
+class IteratorBase:
+    """
+    An Iterator is a wrapper around a pointer, and must define the following:
+
+    - a `state` property that returns a `ctypes.c_void_p` object, representing
+      a pointer to some data.
+    - an `advance` (static) method that receives the state pointer and performs
+      an action that advances the pointer by the offset `distance`
+      (returns nothing).
+    - a `dereference` (static) method that dereferences the state pointer
+      and returns a value.
+    """
```

Comment:
Maybe add

Note: `advance` and `dereference` exist exclusively for compilation with numba. They are not meant to be called from Python.

Reply:
Added some clarifying notes.